pkg/receive/multitsdb.go

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package receive
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	terrors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"github.com/thanos-io/thanos/pkg/shipper"
+	"github.com/thanos-io/thanos/pkg/store"
+	"golang.org/x/sync/errgroup"
+)
+
+type MultiTSDB struct {
+	dataDir         string
+	logger          log.Logger
+	reg             prometheus.Registerer
+	tsdbCfg         *tsdb.Options
+	tenantLabelName string
+	labels          labels.Labels
+	bucket          objstore.Bucket
+
+	mtx     *sync.RWMutex
+	tenants map[string]*tenant
+}
+
+type tenant struct {
+	tsdbCfg *tsdb.Options
+
+	readyS *tsdb.ReadyStorage
```

Comment:
TBH all of those could be be embedded types, but not a big deal

Reply:
so you would not need special methods for those like `storage`

Reply:
Unless all of them have the same method like `close` ;p

Reply:
We need to protect reading them with the mutex, and we would have public functions that don't really make sense in their mix. I think the point is that it's encapsulated, and potentially we can expand this encapsulation to have the tenant object take care of writing/appending directly to rate limit etc.